scipy/optimize/_qap.py

```diff
@@ -0,0 +1,784 @@
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+
+    # OPTIMIZATION WHILE LOOP BEGINS
+    for n_iter in range(maxiter):
+        # computing the gradient of f(P) = -tr(APB^tP^t)
+        grad_fp = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+        # run hungarian algorithm on gradient(f(P))
+        rows, cols = linear_sum_assignment(obj_func_scalar * grad_fp)
+        Q = np.zeros((n_unseed, n_unseed))
+        Q[rows, cols] = 1  # initialize search direction matrix Q
+
+        def f(x):  # computing the original optimization function
+            xP1xQ = x * P + (1 - x) * Q
+            # Sums below are np.trace(A11.T @ B11)
+            # + np.trace(xP1xQ.T @ A21 @ B21.T)
+            # + np.trace(xP1xQ.T @ A12.T @ B12)
+            # + np.trace(A22.T @ xP1xQ @ B22 @ xP1xQ.T)
+            # This is more efficient, but can we do even better?
+            return obj_func_scalar * (
+                (A11 * B11).sum()
+                + (xP1xQ.T @ A21 * B21).sum()
+                + (xP1xQ.T @ A12.T * B12.T).sum()
+                + ((xP1xQ.T @ A22) * (B22 @ xP1xQ.T)).sum()
+            )
+
+        # computing the step size
+        alpha = minimize_scalar(f, bounds=(0, 1), method="bounded").x
+        P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+        if np.linalg.norm(P - P_i1) < tol:
+            break
+        P = P_i1
+    # end of FW optimization loop
+
+    # Project onto the set of permutation matrices
+    row, col = linear_sum_assignment(-P)
+    perm = np.concatenate((np.arange(n_seeds), col + n_seeds))
+
+    unshuffled_perm = np.zeros(n, dtype=int)
+    unshuffled_perm[perm_A] = perm_B[perm]
+
+    score = _calc_score(A, B, unshuffled_perm)
+
+    res = {"col_ind": unshuffled_perm, "fun": score, "nit": n_iter+1}
+
+    return OptimizeResult(res)
+
+
+def _check_init_input(init_J, n):
+    row_sum = np.sum(init_J, axis=0)
+    col_sum = np.sum(init_J, axis=1)
+    tol = 1e-3
+    msg = None
+    if init_J.shape != (n, n):
+        msg = "`init_J` matrix must have same shape as A and B"
+    elif ((~np.isclose(row_sum, np.ones(n), atol=tol)).any() or
+          (~np.isclose(col_sum, np.ones(n), atol=tol)).any() or
+          (init_J < 0).any()):
+        msg = "`init_J` matrix must be doubly stochastic"
+    if msg is not None:
+        raise ValueError(msg)
+
+
+def _split_matrix(X, n):
+    # definitions according to Seeded Graph Matching [2].
+    upper, lower = X[:n], X[n:]
+    return upper[:, :n], upper[:, n:], lower[:, :n], lower[:, n:]
+
+
+def _doubly_stochastic(P, tol=1e-3):
```

Comment:
Can we call this `_sinkhorn_knopp`?
Also add a docstring and a reference.

Reply:
`_doubly_stochastic` is a more readable name, but a reference is appropriate. I don't think a full docstring is needed as it's quite clear what the inputs and outputs are. Perhaps this is motivated by the talk on the mailing list of making a user-facing function for this algorithm? This is still private, so let's save that for another PR.

Reply:
OK. I appreciate the extended comments and the reference.
Resolving.